### PR TITLE
CTCTRADERS-129, created new pages for Simplified journey

### PR DESCRIPTION
**Pull request by kenneth-rayner**
1. Authorisation location code
2. consignee name 
3. eori confirmation
4. eori number
5. consignee address, added Navigation through journey  and dynamic content(ticket 302 in backlog has content)

### CODE REVIEW

File: app/forms/AuthorisedLocationFormProvider.scala [hunk 1]
```diff
@@ -23,9 +23,14 @@ import play.api.data.Form
 
 class AuthorisedLocationFormProvider @Inject() extends Mappings {
```
Comment (chris-ianson) [inline app/forms/AuthorisedLocationFormProvider.scala:24]: Where's the Spec for this?

File: app/navigation/Navigator.scala [hunk 2]
```diff
@@ -38,6 +38,11 @@ class Navigator @Inject()() {
   private val normalRoutes: PartialFunction[Page, UserAnswers => Option[Call]] = {
     case MovementReferenceNumberPage => ua => Some(routes.GoodsLocationController.onPageLoad(ua.id, NormalMode))
     case GoodsLocationPage => goodsLocationPageRoutes
+    case AuthorisedLocationPage => ua => Some(routes.ConsigneeNameController.onPageLoad(ua.id, NormalMode))
+    case ConsigneeNamePage => ua => Some(routes.EoriConfirmationController.onPageLoad(ua.id, NormalMode))
+    case EoriConfirmationPage => eoriConfirmationRoutes
+    case EoriNumberPage => ua => Some(routes.ConsigneeAddressController.onPageLoad(ua.id, NormalMode))
+    case ConsigneeAddressPage => ua => Some(routes.IncidentOnRouteController.onPageLoad(ua.id, NormalMode))
```
Comment (chris-ianson) [inline app/navigation/Navigator.scala:45]: Do these have tests in the navigator?

File: app/pages/TraderAddressPage.scala [hunk 3]
```diff
@@ -16,12 +16,12 @@
...
+import models.Address
 import play.api.libs.json.JsPath
 
-case object TraderAddressPage extends QuestionPage[TraderAddress] {
+case object TraderAddressPage extends QuestionPage[Address] {
 
   override def path: JsPath = JsPath \ toString
 
-  override def toString: String = "traderAddress"
+  override def toString: String = "Address"
```
Comment (chris-ianson) [inline app/pages/TraderAddressPage.scala:26]: Make this lowercase

File: conf/application.conf [hunk 4]
```diff
@@ -139,4 +139,5 @@ nunjucks {
 
 feature-toggles {
   arrivalRejection = true
+  consigneeLocation = true
```
Comment (chris-ianson) [inline conf/application.conf:142]: Comment above about renaming this

File: conf/messages.en [hunk 5]
```diff
@@ -425,3 +426,38 @@ confirmRemoveEvent.error.required = Select yes if you are sure you want to remov
...
+consigneeAddress.title = What is the address of the authorised consignee
+consigneeAddress.heading =  What is the address for {0}?
+consigneeAddress.checkYourAnswersLabel =  What is the address for {0}?
+consigneeAddress.error.required = Enter {0}'s Address
+consigneeAddress.error.length = The address for {0} must be 100 characters or less
+
+consigneeAddress.buildingAndStreet = Building and street
+consigneeAddress.city = Town or city
+consigneeAddress.postcode = Postcode
+traderAddress.checkYourAnswersLabel = Address
```
Comment (chris-ianson) [inline conf/messages.en:456]: Is this key correct? All the others relate to consigneeAddress

File: test/controllers/GoodsLocationControllerSpec.scala [hunk 6]
```diff
@@ -116,17 +109,18 @@ class GoodsLocationControllerSpec extends SpecBase with MockitoSugar with Nunjuc
       application.stop()
     }
 
-    "must redirect to the next page when valid data is submitted" in {
+    "must redirect to the correct page when valid data is submitted and consignee location toggle is false" in {
```
Comment (chris-ianson) [inline test/controllers/GoodsLocationControllerSpec.scala:112]: Great that you're testing the toggle. When you update the toggle name, just be sure to update the spec descriptions here

File: test/controllers/GoodsLocationControllerSpec.scala [hunk 7]
```diff
@@ -143,6 +137,60 @@ class GoodsLocationControllerSpec extends SpecBase with MockitoSugar with Nunjuc
       application.stop()
     }
 
+    "must redirect to the correct page for Border Force Office when valid data is submitted and consignee location toggle is true" in {
```
Comment (chris-ianson) [inline test/controllers/GoodsLocationControllerSpec.scala:140]: 👍

File: app/config/FrontendAppConfig.scala [hunk 8]
```diff
@@ -47,7 +47,8 @@ class FrontendAppConfig @Inject()(configuration: Configuration) {
   lazy val destinationUrl: String   = configuration.get[Service]("microservice.services.destination").baseUrl
   lazy val referenceDataUrl: String = configuration.get[Service]("microservice.services.referenceData").baseUrl
 
-  lazy val featureToggleArrivalRejection: Boolean = configuration.getOptional[Boolean]("feature-toggles.arrivalRejection").getOrElse(false)
+  lazy val featureToggleArrivalRejection: Boolean  = configuration.getOptional[Boolean]("feature-toggles.arrivalRejection").getOrElse(false)
+  lazy val featureToggleConsigneeLocation: Boolean = configuration.getOptional[Boolean]("feature-toggles.consigneeLocation").getOrElse(false)
```
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/config/FrontendAppConfig.scala:51]: Changed to featureToggleSimplifiedJourney

File: app/forms/AuthorisedLocationFormProvider.scala [hunk 9]
```diff
@@ -23,9 +23,14 @@ import play.api.data.Form
...
 
+  val alRegex = "^[a-zA-Z0-9]*$"
```
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/forms/AuthorisedLocationFormProvider.scala:26]: In the ticket it states 'Authorised location code must only contain letters a to z or numbers 0 to 9'

File: conf/application.conf (hunk 4 above)
Comment (kenneth-rayner) [reply conf/application.conf:142]: Already done

File: conf/messages.en (hunk 5 above)
Comment (kenneth-rayner) [reply conf/messages.en:456]: Done

File: app/config/FrontendAppConfig.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/config/FrontendAppConfig.scala:51]: Done

File: app/forms/AuthorisedLocationFormProvider.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (kenneth-rayner) [reply app/forms/AuthorisedLocationFormProvider.scala:26]: ticket states 
Authorised location code must only contain letters a to z or numbers 0 to 9

File: app/pages/TraderAddressPage.scala (hunk 3 above)
Comment (kenneth-rayner) [reply app/pages/TraderAddressPage.scala:26]: done

File: conf/application.conf (hunk 4 above)
Comment (kenneth-rayner) [reply conf/application.conf:142]: done

File: conf/messages.en (hunk 5 above)
Comment (kenneth-rayner) [reply conf/messages.en:456]: done

File: test/controllers/GoodsLocationControllerSpec.scala (hunk 6 above)
Comment (kenneth-rayner) [reply test/controllers/GoodsLocationControllerSpec.scala:112]: done

File: test/controllers/GoodsLocationControllerSpec.scala (hunk 7 above)
Comment (kenneth-rayner) [reply test/controllers/GoodsLocationControllerSpec.scala:140]: done

File: app/forms/AuthorisedLocationFormProvider.scala (hunk 1 above)
Comment (kenneth-rayner) [reply app/forms/AuthorisedLocationFormProvider.scala:24]: done

File: app/forms/AuthorisedLocationFormProvider.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (chris-ianson) [reply app/forms/AuthorisedLocationFormProvider.scala]: Nice one, thanks Ken (the ticket had that much info?!)

File: app/pages/TraderAddressPage.scala (hunk 3 above)
Comment (chris-ianson) [reply app/pages/TraderAddressPage.scala]: 👍

File: conf/application.conf (hunk 4 above)
Comment (chris-ianson) [reply conf/application.conf]: Thanks. Keep in mind when this goes in master we'll want the toggle to false

File: test/navigation/NavigatorSpec.scala [hunk 10]
```diff
@@ -67,34 +67,93 @@ class NavigatorSpec extends SpecBase with ScalaCheckPropertyChecks with Generato
 
       }
 
-      "must go from 'goods location' to  'customs approved location' when user chooses 'Border Force Office'" in {
+      "must go from 'goods location' to  'customs approved location' when user chooses 'Border Force Office' when Simplified Journey toggle is on or off" in {
         forAll(arbitrary[UserAnswers]) {
           answers =>
-            val updatedAnswers = answers.set(GoodsLocationPage, GoodsLocation.BorderForceOffice).success.value
+            if (frontendAppConfig.featureToggleSimplifiedJourney) {
```
Comment (chris-ianson) [inline test/navigation/NavigatorSpec.scala:73]: Have a look at manage frontend to see how we have the toggle in ViewMovementSpec. You need to override the app on setup and set the value you're testing for. Have a look and we'll go over this tomorrow. The problem with having an if condition to lookup the config value is you're only ever testing half the implementation.

File: test/forms/AuthorisedLocationFormProviderSpec.scala [hunk 11]
```diff
@@ -17,37 +17,42 @@
...
-      lengthError = FormError(fieldName, lengthKey, Seq(maxLength))
-    )
-
     behave like mandatoryField(
       form,
       fieldName,
       requiredError = FormError(fieldName, requiredKey)
     )
   }
+  "must not bind strings that do not match regex" in {
```
Comment (chris-ianson) [inline test/forms/AuthorisedLocationFormProviderSpec.scala:46]: I like this 👍

File: test/navigation/NavigatorSpec.scala (hunk 10 above)
Comment (chris-ianson) [reply test/navigation/NavigatorSpec.scala:73]: Having a think about this I'm not sure if we need the toogle here. We're only using the toggle in the controller aren't we? And you have the toggle tested in the controller. We'll go over it tomorrow.

File: app/controllers/CheckYourAnswersController.scala [hunk 12]
```diff
@@ -90,19 +101,33 @@ class CheckYourAnswersController @Inject()(override val messagesApi: MessagesApi
...
     )
+    val consigneeDetails = Section(
+      msg"checkYourAnswers.section.consigneeDetails",
+      Seq(
+        helper.consigneeName,
+        helper.eoriConfirmation(eori),
+        userAnswers.get(EoriConfirmationPage) match {
+          case Some(false) => helper.eoriNumber
+          case _           => None
+        },
```
Comment (chris-ianson) [inline app/controllers/CheckYourAnswersController.scala:112]: Why do you have helper.eoriNumber wrapped in a match? It does this anyway in eoriNumber def

File: app/controllers/CheckYourAnswersController.scala [hunk 13]
```diff
@@ -73,15 +75,24 @@ class CheckYourAnswersController @Inject()(override val messagesApi: MessagesApi
...
+            msg"checkYourAnswers.section.goodsLocation",
+            Seq(helper.goodsLocation, helper.authorisedLocation, helper.customsSubPlace, helper.presentationOffice).flatten
+          )
+        case _ =>
+          Section(
+            msg"checkYourAnswers.section.goodsLocation",
+            Seq(helper.goodsLocation, helper.authorisedLocation).flatten
+          )
+      }
+    )
```
Comment (chris-ianson) [inline app/controllers/CheckYourAnswersController.scala:95]: The only difference I see between normal and simplified is the additional customsSubPlace and presentationOffice. To reduce some of the duplication have the match only return a Seq(helper.customsSubPlace, helper.presentationOffice) if the user answer is BorderForceOffice. The append this Seq to Seq(helper.goodsLocation, helper.authorisedLocation) Have a go at doing this and if you get stuck give me a shout.

File: app/controllers/GoodsLocationController.scala [hunk 14]
```diff
@@ -83,7 +86,18 @@ class GoodsLocationController @Inject()(override val messagesApi: MessagesApi,
             for {
               updatedAnswers <- Future.fromTry(request.userAnswers.getOrElse(UserAnswers(mrn)).set(GoodsLocationPage, value))
               _              <- sessionRepository.set(updatedAnswers)
-            } yield Redirect(navigator.nextPage(GoodsLocationPage, mode, updatedAnswers))
+            } yield
+              if (frontendAppConfig.featureToggleSimplifiedJourney) {
```
Comment (chris-ianson) [inline app/controllers/GoodsLocationController.scala:90]: Could we put the feature toggle into the AuthorisedConsigneesLocation match block please? BorderForceOffice always goes to the same place.

File: app/navigation/Navigator.scala [hunk 15]
```diff
@@ -79,22 +89,33 @@ class Navigator @Inject()() {
...
   }
 
+
+  def eoriConfirmationRoutes(ua: UserAnswers, mode:Mode) =
+    ((ua.get(EoriConfirmationPage)),mode) match {
+      case (Some(true), NormalMode) => routes.ConsigneeAddressController.onPageLoad(ua.id, mode)
+      case (Some(true), CheckMode) => routes.CheckYourAnswersController.onPageLoad(ua.id)
+      case _ => routes.EoriNumberController.onPageLoad(ua.id, mode)
+//      case (Some(false), CheckMode) => routes.EoriNumberController.onPageLoad(ua.id, CheckMode)
+//      case _ => routes.EoriNumberController.onPageLoad(ua.id, NormalMode)
```
Comment (chris-ianson) [inline app/navigation/Navigator.scala:107]: Remove comments if they're not needed

File: conf/application.conf [hunk 16]
```diff
@@ -139,4 +139,5 @@ nunjucks {
 
 feature-toggles {
   arrivalRejection = true
+  simplifiedJourney = true
```
Comment (chris-ianson) [inline conf/application.conf:142]: Does this need to be false when it's merged. Need to speak to Paul about this and understand how toggles are tested (if they are)

File: app/controllers/CheckYourAnswersController.scala (hunk 12 above)
Comment (kenneth-rayner) [reply app/controllers/CheckYourAnswersController.scala:112]: Amended

File: app/forms/EoriConfirmationFormProvider.scala [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+import javax.inject.Inject
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+class EoriConfirmationFormProvider @Inject() extends Mappings {
+
+  def apply(traderName: String): Form[Boolean] =
+    Form(
+      "value" -> boolean(requiredKey = "eoriConfirmation.error.required", args = Seq(traderName))
```
Comment (PSouthern) [inline app/forms/EoriConfirmationFormProvider.scala:28]: When this error is thrown the page heading isn't being displayed correctly.

`What is the EORI number for org.mozilla.javascript.Undefined@0?` is shown

File: app/navigation/Navigator.scala (hunk 15 above)
Comment (kenneth-rayner) [reply app/navigation/Navigator.scala:107]: Done

File: app/controllers/GoodsLocationController.scala (hunk 14 above)
Comment (kenneth-rayner) [reply app/controllers/GoodsLocationController.scala:90]: done

File: app/forms/EoriConfirmationFormProvider.scala (hunk 17 above)
Comment (kenneth-rayner) [reply app/forms/EoriConfirmationFormProvider.scala:28]: Done

File: app/controllers/CheckYourAnswersController.scala (hunk 13 above)
Comment (kenneth-rayner) [reply app/controllers/CheckYourAnswersController.scala:95]: done

File: conf/application.conf (hunk 16 above)
Comment (kenneth-rayner) [reply conf/application.conf:142]: left as true for now

File: app/pages/GoodsLocationPage.scala [hunk 18]
```diff
@@ -29,8 +29,24 @@ case object GoodsLocationPage extends QuestionPage[GoodsLocation] {
 
   override def cleanup(value: Option[GoodsLocation], userAnswers: UserAnswers): Try[UserAnswers] =
     value match {
-      case Some(GoodsLocation.AuthorisedConsigneesLocation) => userAnswers.remove(CustomsSubPlacePage)
-      case Some(GoodsLocation.BorderForceOffice)            => userAnswers.remove(AuthorisedLocationPage)
-      case None                                             => super.cleanup(value, userAnswers)
+      case Some(GoodsLocation.AuthorisedConsigneesLocation) =>
```
Comment (chris-ianson) [inline app/pages/GoodsLocationPage.scala:32]: Looks good that we're removing user answers. Could we add to the GoodsLocationPageSpec to make sure all relevant values are removed? (don't mind if it's multiple assertions in each of the specs)

File: test/navigation/NavigatorSpec.scala [hunk 19]
```diff
@@ -88,38 +88,46 @@ class NavigatorSpec extends SpecBase with ScalaCheckPropertyChecks with Generato
           answers =>
             navigator
               .nextPage(ConsigneeNamePage, NormalMode, answers)
-              .mustBe(routes.EoriConfirmationController.onPageLoad(answers.id, NormalMode))
+              .mustBe(routes.ConsigneeEoriConfirmationController.onPageLoad(answers.id, NormalMode))
         }
       }
+
       "must go from'eoriConfirmation page to eoriNumber page when 'No' is selcted" in {
```
Comment (chris-ianson) [inline test/navigation/NavigatorSpec.scala:95]: typo

File: test/navigation/NavigatorSpec.scala [hunk 20]
```diff
@@ -88,38 +88,46 @@ class NavigatorSpec extends SpecBase with ScalaCheckPropertyChecks with Generato
...
 
             navigator
-              .nextPage(EoriConfirmationPage, NormalMode, updatedAnswers)
-              .mustBe(routes.EoriNumberController.onPageLoad(answers.id, NormalMode))
+              .nextPage(ConsigneeEoriConfirmationPage, NormalMode, updatedAnswers)
+              .mustBe(routes.ConsigneeEoriNumberController.onPageLoad(answers.id, NormalMode))
         }
       }
+
       "must go from'eoriConfirmation page to consigneeAddress page when 'Yes' is selcted" in {
```
Comment (chris-ianson) [inline test/navigation/NavigatorSpec.scala:112]: typo

File: test/pages/ConsigneeEoriConfirmationPageSpec.scala [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+              .set(ConsigneeEoriConfirmationPage, false)
+              .success
+              .value
+              .set(ConsigneeEoriNumberPage, eoriNumber)
+              .success
+              .value
+              .set(ConsigneeEoriConfirmationPage, true)
+              .success
+              .value
+            result.get(ConsigneeEoriNumberPage) must not be defined
```
Comment (chris-ianson) [inline test/pages/ConsigneeEoriConfirmationPageSpec.scala:48]: Assert all values are removed here

File: test/pages/ConsigneeEoriConfirmationPageSpec.scala [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+              .set(ConsigneeEoriConfirmationPage, true)
+              .success
+              .value
+              .set(ConsigneeEoriNumberPage, eoriNumber)
+              .success
+              .value
+              .set(ConsigneeEoriConfirmationPage, false)
+              .success
+              .value
+            result.get(ConsigneeEoriNumberPage) mustBe defined
```
Comment (chris-ianson) [inline test/pages/ConsigneeEoriConfirmationPageSpec.scala:64]: Assert all values are removed here

File: app/navigation/Navigator.scala [hunk 23]
```diff
@@ -131,6 +127,62 @@ class Navigator @Inject()() {
...
+      case (Some(true), CheckMode, _, _)        => Some(routes.CheckYourAnswersController.onPageLoad(ua.id))
+      case (Some(false), CheckMode, Some(_), _) => Some(routes.CheckYourAnswersController.onPageLoad(ua.id))
+      case (Some(false), NormalMode, _, _)      => Some(routes.ConsigneeEoriNumberController.onPageLoad(ua.id, mode))
+      case (Some(false), CheckMode, None, _)    => Some(routes.ConsigneeEoriNumberController.onPageLoad(ua.id, mode))
+      case _                                    => None
+    }
+
+  private def consigneeEoriNumberRoute(mode: Mode)(ua: UserAnswers) =
+    (mode, ua.get(ConsigneeAddressPage)) match {
+      case (NormalMode, _)   => Some(routes.ConsigneeAddressController.onPageLoad(ua.id, mode))
```
Comment (chris-ianson) [inline app/navigation/Navigator.scala:150]: Look at refactoring this

File: app/pages/GoodsLocationPage.scala (hunk 18 above)
Comment (kenneth-rayner) [reply app/pages/GoodsLocationPage.scala:32]: Done

File: test/navigation/NavigatorSpec.scala (hunk 19 above)
Comment (kenneth-rayner) [reply test/navigation/NavigatorSpec.scala:95]: Done

File: test/navigation/NavigatorSpec.scala (hunk 20 above)
Comment (kenneth-rayner) [reply test/navigation/NavigatorSpec.scala:112]: Done

File: app/navigation/Navigator.scala (hunk 23 above)
Comment (kenneth-rayner) [reply app/navigation/Navigator.scala:150]: Done

File: test/pages/ConsigneeEoriConfirmationPageSpec.scala (hunk 21 above)
Comment (kenneth-rayner) [reply test/pages/ConsigneeEoriConfirmationPageSpec.scala:48]: no

Comment (kenneth-rayner) [reply test/pages/ConsigneeEoriConfirmationPageSpec.scala:48]: no

File: test/pages/ConsigneeEoriConfirmationPageSpec.scala (hunk 22 above)
Comment (kenneth-rayner) [reply test/pages/ConsigneeEoriConfirmationPageSpec.scala:64]: no

**Pull request merged**